File: src/cardTransformationUtils.ts
import { PartialScryfallCard, CardFace, Card } from "./types";

function cardType(card: PartialScryfallCard | CardFace): string | null {
  if ("type_line" in card) {
    return card["type_line"].toLowerCase().includes("land") ? "Land" : "Nonland";
  } else {
    return null;
  }
}

function producedMana(card: PartialScryfallCard): string | null {
  if (cardType(card) !== "Land") {
    return null;
  }

  else if ("card_faces" in card) {
    const colors: string[] = [];
    for (const face of card["card_faces"]) {
      const match = face.oracle_text?.match(/^(.*?|\n)?\{T\}:.*?[Aa]dd(.+?\{(([A-Z]\/?)+?)\})+/);
      if (match) {
        const newColors = face.oracle_text!
          .substring(match.index!, match.index! + match[0].length)
          .match(/\{([^T]+)\}/g)
          ?.map(mana => mana.substring(1, mana.length - 1))
          .flatMap(mana => mana.split("/")) || [];
        colors.push(...newColors);
      }
    }
    return [...new Set(colors)].join(",") || null;
  }

  else if ("produced_mana" in card) {
    return card["produced_mana"]?.join(",");
  }

  else {
    return null;
  }
}

function tapLand(card: PartialScryfallCard): boolean | null {
  if (cardType(card) !== "Land") {
    return null;
  }

  const hasTapLandPhrase = (c: PartialScryfallCard | CardFace): boolean => {
    return c.oracle_text ? 
      RegExp(`(NAME enters the battlefield tapped.(?!unless))|(Hideaway)|(NAME enters tapped)`)
        .test(c.oracle_text.replace(c.name!, "NAME")) : 
      false;
  };
  
  let isTapLand: boolean | null = null;

  if ("card_faces" in card) {
    for (const face of card["card_faces"]!) {
      if (cardType(face) === "Land") {
        isTapLand = hasTapLandPhrase(face);
      }
    }
  } else {
    isTapLand = hasTapLandPhrase(card);
  }
    
  return isTapLand;
}

function manaCost(card: PartialScryfallCard): string | null {
  let cost: string | null = null;
  if ("card_faces" in card) {
    for (const face of card["card_faces"]!) {
      if ("mana_cost" in face && face["mana_cost"]) {
        cost = face["mana_cost"];
      }
    }
  } else if ("mana_cost" in card) {
    cost = card["mana_cost"];
  }
  return cost;
}


type TransformationFunction = (card: PartialScryfallCard) => string | boolean | null;

const transformationLookup: Record<string, TransformationFunction> = {
  "mana_cost": manaCost,
  "types": (card: PartialScryfallCard): string | null => "type_line" in card ? card["type_line"] : null,
  "type": cardType as TransformationFunction,
  "producible_mana_colors": producedMana,
  "tap_land": tapLand
};

export function scryfallToCard(partialScryfallCard: PartialScryfallCard): Card {
  const card: Partial<Card> = { name: partialScryfallCard.name };

  for (const key in transformationLookup) {
    const transformedValue = transformationLookup[key](partialScryfallCard);
    if (transformedValue !== null) {
      (card as any)[key] = transformedValue;
    }
  }

  return card as Card;
}

File: scripts/card_transformation_utils.py
import re

def card_type(card):
  if "type_line" in card:
    return "Land" if "land" in card["type_line"].lower() else "Nonland"
  else:
    return None

def produced_mana(card):
  if card_type(card) != "Land":
    return None

  elif "card_faces" in card:
    colors = []
    for face in card["card_faces"]:
      match = re.match(r"^(.*?|\n)?\{T\}:.*?[Aa]dd(.+?\{(([A-Z]\/?)+?)\})+", face["oracle_text"])
      if match:
        new_colors = sum([mana.split("/") for mana in re.findall(r"\{([^T]+)\}", face["oracle_text"][match.span()[0]:match.span()[1]])], [])
        colors += new_colors
    return ",".join(list(set(colors))) or None

  elif "produced_mana" in card:
    return ",".join(card["produced_mana"])

  else:
    return None

def tap_land(card):
  if card_type(card) != "Land":
    return None

  has_tap_land_phrase = lambda c: re.search(f"(NAME enters the battlefield tapped.(?!unless))|(Hideaway)|(NAME enters tapped)", c["oracle_text"].replace(c["name"], "NAME")) != None
  is_tap_land = None

  if "card_faces" in card:
    for face in card["card_faces"]:
      if card_type(face) == "Land":
        is_tap_land = has_tap_land_phrase(face)
  else:
    is_tap_land = has_tap_land_phrase(card)
    

  return is_tap_land

def mana_cost(card):
  cost = None
  if "card_faces" in card:
    for face in card["card_faces"]:
      if "mana_cost" in face and face["mana_cost"]:
        cost = face["mana_cost"]
  elif "mana_cost" in card:
    cost = card["mana_cost"]
  return cost


# methods for transforming raw card data to the specific felids for the format that the probability calculator uses
transformation_lookup = {
  "mana_cost": mana_cost,
  "types": lambda card: card["type_line"] if "type_line" in card else None,
  "type": card_type,
  "producible_mana_colors": produced_mana,
  "tap_land": tap_land
}

# function that generates a card in the format the probability calculator uses
def scryfall_to_local_format(scryfall_card):
  card = { "name": scryfall_card["name"] }

  for key in transformation_lookup:
    transformed_value = transformation_lookup[key](scryfall_card)
    if transformed_value != None:
      card[key] = transformed_value

  return card

__all__ = scryfall_to_local_format